### PR TITLE
Switch to typescript, bring repo up to date

### PR DESCRIPTION
**Pull request by jamesrollings**
Switch to typescript
Fix bug with file still being processed even if no mapping exists
Project is now async

### CODE REVIEW

File: src/watchFiles.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
+import path from 'path';
+import chokidar from 'chokidar';
+import chalk from 'chalk';
+
+dotenv.config();
 
 const folderName = 'Downloads';
-const basePath = path.resolve(process.env.HOMEDRIVE, process.env.HOMEPATH);
-const folderToWatch = path.resolve(basePath, folderName);
+const basePath: string = path.resolve(process.env.HOMEDRIVE!, process.env.HOMEPATH!);
```
Comment (connorjburton) [inline src/watchFiles.ts]: These env vars could be null, don't ignore them, check for them at the start of the script and handle it (exit?)

File: src/watchFiles.ts [hunk 2]
```diff
@@ -1,14 +1,16 @@
-require('dotenv').config();
-const fs = require('fs');
-const path = require('path');
-const chokidar = require('chokidar');
-const chalk = require('chalk');
+import dotenv from 'dotenv';
+import fs from 'fs-extra';
```
Comment (connorjburton) [inline src/watchFiles.ts]: You shouldn't need this package (assuming you added it for async support), node has it's own async versions of fs methods. https://nodejs.org/api/fs.html#promises-api

File: src/watchFiles.ts [hunk 3]
```diff
@@ -1,14 +1,16 @@
...
+import chokidar from 'chokidar';
+import chalk from 'chalk';
+
+dotenv.config();
 
 const folderName = 'Downloads';
-const basePath = path.resolve(process.env.HOMEDRIVE, process.env.HOMEPATH);
-const folderToWatch = path.resolve(basePath, folderName);
+const basePath: string = path.resolve(process.env.HOMEDRIVE!, process.env.HOMEPATH!);
+const folderToWatch: string = path.resolve(basePath, folderName);
```
Comment (connorjburton) [inline src/watchFiles.ts]: You probably don't need to be super specific here, as long as `path.resolve` always returns a stirng then let TypeScript type this automatically for you.

File: src/watchFiles.ts [hunk 4]
```diff
@@ -1,14 +1,16 @@
...
 
-const extensions = {
+const extensions: {[path: string] : string[]} = {
```
Comment (connorjburton) [inline src/watchFiles.ts]: It's useful to split these types out into their own type rather than doing it inline here

File: dist/watchFiles.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
+"use strict";
```
Comment (connorjburton) [inline dist/watchFiles.js]: Don't include `dist` files in source control, remove this and gitignore this folder.

**Review by connorjburton — CHANGES_REQUESTED** [review]

File: src/watchFiles.ts (hunk 2 above)
Comment (jamesrollings) [reply src/watchFiles.ts]: `fs-extra` is a drop in replacement for `fs` and/or `fs/promises`, it provides a `pathExists` function which is using `fs.access` along with some other checks, it wraps it all up into one call

Comment (connorjburton) [reply src/watchFiles.ts]: [I wouldn't use it, `pathExists` is literally just `access` wrapped in a promise](https://github.com/jprichardson/node-fs-extra/blob/master/lib/path-exists/index.js), not worth adding a dependency to do that, write it yourself. You should have a very good reason to add any dependency, this doesn't seem one.

**Review by connorjburton — APPROVED** [review]
Good job

**Pull request merged**